source/Timer.cpp

```diff
@@ -0,0 +1,376 @@
+				return "on reset";
+			default:
+				return "unknown trigger";
+		}
+	}
+}
+
+
+
+Timer::Timer(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Timer::Load(const DataNode &node)
+{
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "time" && child.Size() > 1)
+		{
+			waitTime = child.Value(1);
+			if(child.Size() > 2)
+				randomWaitTime = child.Value(2);
+		}
+		else if(child.Token(0) == "idle")
+		{
+			requireIdle = true;
+			// We square the max speed value here, so it can be conveniently
+			// compared to the flagship's squared velocity length below.
+			if(child.Size() > 1)
+				idleMaxSpeed = child.Value(1) * child.Value(1);
+		}
+		else if(child.Token(0) == "peaceful")
+			requirePeaceful = true;
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity" && child.Size() > 1)
+		{
+			proximityCenter = GameData::Planets().Get(child.Token(1));
+		}
+		else if(child.Token(0) == "proximity")
+		{
+			proximityCenters.Load(child);
+		}
+		else if(child.Token(0) == "proximity settings" && child.Size() > 1)
+		{
+			proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset" && child.Size() > 1)
+		{
+			const string &value = child.Token(1);
+			if(value == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(value == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(value == "none")
+				resetCondition = ResetCondition::NONE;
+			else if(value == "pause")
+				resetCondition = ResetCondition::PAUSE;
+			else
+				child.PrintTrace("Skipping unrecognized attribute:");
+		}
+		else if(child.Token(0) == "repeat reset")
+			repeatReset = true;
+		else if(child.Token(0) == "reset fired")
+			resetFired = true;
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
+			actions[TimerTrigger::TIMEUP].Load(child);
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
+			actions[TimerTrigger::RESET].Load(child);
+		else
+			child.PrintTrace("Skipping unrecognized attribute:");
+	}
+}
+
+
+
+// Note: the Save() function can assume this is an instantiated Timer, not a template,
+// so the time to wait will be saved fully calculated, and with any elapsed time subtracted.
+void Timer::Save(DataWriter &out) const
+{
+	// If this Timer should no longer appear in-game, don't serialize it.
+	if(isComplete)
+		return;
+
+	int64_t timeRemaining = waitTime - timeElapsed;
+	out.Write("timer");
+	out.BeginChild();
+	{
+		out.Write("time", timeRemaining);
```

Comment:
We need to separately save the time elapsed and the wait time instead of calculating a time remaining and saving that, as otherwise when we save the game we have no idea what to reset the timer to if it need reset, or it'll reset to the wrong value.

Reply:
OK. Adding this.